dorado/lagrangian_walker.py

```diff
@@ -38,11 +39,73 @@ def random_pick_seed(choices, probs=None):
     return choices[idx]
 
 
+def make_weight(Particles):
+    """Make an array with the routing weights."""
+    # init the weight array
+    L, W = np.shape(Particles.stage)
+    Particles.weight = np.zeros((L, W, 9))
+    # do weighting calculation for each cell
+    print('Calculating routing weights ...')
+    for i in tqdm(list(range(1, L-1)), ascii=True):
+        for j in list(range(1, W-1)):
```

Comment:
So, looking at this code, I see that the main structure of the weight computation hasn't changed. We're still constructing small sub-arrays at each index, doing a few quick operations (max, cleaning, multiplications, summing), and saving the 9 resulting weights in a `weight` array. Surprised we hadn't tried this sooner given how much better it performs.

It looks like the key reason the runtime scales better in this model isn't anything about how this computation works, it's how many times we do it. Originally, we performed these operations locally once for every particle at every iteration. So, the runtime scaled as `Np_tracer * iterations_per_particle`. Now, we perform this once for every cell, so it scales with domain size `(L-2) * (W-2)`. I bet if you checked the example cases you benchmarked, the ratio of these values should give roughly the speedup you observed. 

One thing I wonder, though, is whether we could obtain even faster runtimes by modifying the structure of this computation itself, by switching from local array operations to global. Whatever is causing this function to take so long must be due to the fact that we're repeatedly constructing many sub-arrays in a loop and operating on them, instead of performing a few big global array operations (inside which we'd be taking advantage of all the fast numpy broadcasting stuff). In principal, if we broke up these operations (max, cleaning, multiplication, summing) into a loop over each of the D8 directions, with each being a global matrix operation, instead of a loop over each cell, we could reduce the amount of overhead repeatedly calling these functions. I don't know exactly how that scales, but it'd be the difference between `time(np.max(small array)) * (L-2) * (W-2)` and `time(np.max(big array)) * 9`. Does that scale better? Not sure.

That being said, I think it's very fine to commit these changes now -- the PR looks good, and it's still a significant improvement in runtime. Might be worth investigating if the global method will be better in a future change.

Reply:
Yeah this is a good point. 

It is pretty much just a port of the previous `get_weight()` functionality to doing it upfront for all locations. I agree that the trade-off in time with the methods should just about match what you are saying. 

I'm going to open up an issue with your commentary on this so that we don't forget to implement it in the future.